hypothesis-python/src/hypothesis/internal/conjecture/shrinking/length.py

```diff
@@ -52,50 +53,87 @@ def left_is_better(self, left, right):
+
+            keep = keep & current_subset
+            if len(keep) == len(current_subset):
+                return True
+            to_remove = current_subset - keep
+            if not to_remove.issubset(candidates_for_removal):
+                return False
+            if self.consider([v for i, v in enumerate(initial) if i in keep]):
+                current_subset.intersection_update(keep)
+                return True
+            return False
+
+        # We iterate over the indices in random order. This is because deletions
+        # towards the end are more likely to work, while deletions from the
+        # beginning are more likely to have higher impact. In addition there
+        # tend to be large "dead" regions where nothing can be deleted, and
+        # by proceeding in random order we don't have long gaps in those where
+        # we make no progress.
+        #
+        # Note that this may be strictly more expensive than iterating from
+        # left to right or right to left. The cost of find_integer, say f, is
+        # convex. When deleting n elements starting from the left we pay f(n)
+        # invocations, but when starting from the middle we pay 2 f(n / 2)
+        # > f(n) invocations. In this case we are prioritising making progress
+        # over a possibly strictly lower cost for two reasons: Firstly, when
+        # n is small we just do linear scans anyway so this doesn't actually
+        # matter, and secondly because successfuly deletions will tend to
+        # speed up the test function and thus even when we make more test
+        # function calls we may still win on time.
+        #
+        # It's also very frustrating watching the shrinker repeatedly fail
+        # to delete, so there's a psychological benefit to prioritising
+        # progress over cost.
+        self.random.shuffle(indices)
+        for i in indices:
+            candidates_for_removal &= current_subset
+            if not candidates_for_removal:
+                break
+
+            # We have already processed this index, either because it was bulk
+            # removed or is the end point of a set that was.
+            if i not in candidates_for_removal:
+                continue
+
+            if not consider_set(current_subset - {i}):
+                candidates_for_removal.discard(i)
+                continue
+
+            # We have successfully removed index i. Now try removing the region
+            # surrounding i, as deletable elements are often inside large deletable
+            # regions. Once we've tried this, we remove the end points from our
+            # candidate set, as we have tried deleting them by virtue of trying to
+            # extend the deleted region to that point and failing.
+            to_right = find_integer(
+                lambda n: i + n <= len(initial)
+                and consider_set(current_subset - set(range(i, i + n)))
+            )
```

Comment:
Note that when `n=1`, this tries to remove just `i` again, which is redundant with what we've already tried above.

I *think* the length check in `consider_set` bails us out here, and if so then it's nice that this keeps the indexing simpler. But I haven't properly convinced myself that this is fine.

Reply:
Hmm. That's a good point. Note that because of the underlying caching in the shrinker this is fine though - `consider` will not invoke the underlying predicate again when it's redundant, so although it's a little bit of wasted work it doesn't invoke the potentially expensive test function.

Reply:
Perhaps the thing to do here is to remove the initial check and _always_ run `to_right`, running `to_left` only if `to_right > 0`.

Reply:
Yeah, that sounds like a good approach.

It gets rid of the benign redundant step, preserves the simpler indexing, and gives context for the asymmetry between `to_right` (inclusive of `i`) and `to_left` (exclusive of `i`).